### PR TITLE
Unstake active validators

### PR DESCRIPTION
**Pull request by enriquefynn**
Adds a function in the maintainer to unstake Lamports from validators if there's the need of a rebalance.

There's a threshold (in %) for when to unstake from validators, the conditions for the unstake are:
- If the amount that should be unstaked is above the threshold.
- If there is some validator which needs to be staked and the amount is above the threshold.
- 
As discussed in #427 with @ruuda this PR implements and tests the unstaking logic.

### CODE REVIEW

File: cli/src/maintenance.rs [hunk 1]
```diff
@@ -798,6 +829,58 @@ impl SolidoState {
...
+
+        // Get the maximum that can be unstaked from the stake account.
+        let amount = unstake_amounts[validator_index].min(stake_account.1.balance.total());
+
+        let expected_balance = (validator.entry.effective_stake_balance() - amount)
+            .expect("amount is always <= balance");
+        let amount = if expected_balance > MINIMUM_STAKE_ACCOUNT_BALANCE {
+            Some(amount)
+        } else {
+            let min_amount = (validator.entry.effective_stake_balance()
```
Comment (ruuda) [inline cli/src/maintenance.rs:861]: This should be the amount in the first stake account, not the total effective balance.

File: tests/test_solido.py [hunk 2]
```diff
@@ -211,46 +213,66 @@ def approve_and_execute(
...
-    'tests/.keys/validator-token-account-key.json', st_sol_mint_account
-)
-print(f'> Validator stSol token account: {validator_fee_account}')
+class Validator(NamedTuple):
+    account: TestAccount
+    vote_account: TestAccount
+    fee_account: str
+
+
+def add_validator(keypath_account: str, keypath_vote: str) -> Tuple[Validator, Any]:
```
Comment (ruuda) [inline tests/test_solido.py:235]: A bit more precise than `Any` for json objects is `Dict[str, Any]`.

File: tests/test_solido.py [hunk 3]
```diff
@@ -484,6 +506,45 @@ def deposit(lamports: int, expect_created_token_account: bool = False) -> None:
...
+)
+
+del result['UnstakeFromActiveValidator']['from_stake_account']
+del result['UnstakeFromActiveValidator']['to_unstake_account']
+expected_result = {
+    'UnstakeFromActiveValidator': {
+        'validator_vote_account': validator.vote_account.pubkey,
+        'from_stake_seed': 0,
+        'to_unstake_seed': 0,
+        'amount': 1500000000,
```
Comment (ruuda) [inline tests/test_solido.py:544]: You can use numeric underscores in Python to make this large number more readable, i.e. `1_500_000_000`.

File: tests/test_solido.py [hunk 4]
```diff
@@ -523,16 +588,58 @@ def deposit(lamports: int, expect_created_token_account: bool = False) -> None:
...
 )
+
+del result['StakeDeposit']['stake_account']
 expected_result = {
     'StakeDeposit': {
-        'validator_vote_account': validator_vote_account.pubkey,
-        # We have 1.0 SOL from the true deposit, and 1.0 donated.
-        'amount_lamports': int(2.0e9),
+        'validator_vote_account': validator_1.vote_account.pubkey,
+        'amount_lamports': 2_050_250_000,
```
Comment (ruuda) [inline tests/test_solido.py:596]: How did we get to this number? It looks a bit arbitrary to me, can you add a comment about why we should get this particular amount? (E.g. deposited X and Y = Z and we divide that over n validators.)

**Review by ruuda — COMMENTED** [review]
This looks mostly good to me! My main comment is that we should unstake from the validator furthest above its target, not just from the first one.

Also, we were planning to avoid any stake/unstake operations until the last hour of the epoch, will that be done in a follow-up PR? (Maybe we should also have a flag to disable this behavior, because it will make testing more difficult.)

File: cli/src/maintenance.rs [hunk 5]
```diff
@@ -95,21 +95,45 @@ pub enum MaintenanceOutput {
...
+    #[serde(serialize_with = "serialize_b58")]
+    from_stake_account: Pubkey,
+    #[serde(serialize_with = "serialize_b58")]
+    to_unstake_account: Pubkey,
+    from_stake_seed: u64,
+    to_unstake_seed: u64,
+    amount: Lamports,
+}
+
+fn print_unstake(f: &mut fmt::Formatter, unstake: &Unstake) -> fmt::Result {
```
[the remark this answers is not in the document]
Comment (enriquefynn) [reply cli/src/maintenance.rs:119]: Great, it's better. Similar to the function below it.

File: program/src/token.rs [hunk 6]
```diff
@@ -22,12 +22,19 @@ use std::{
...
+#[derive(Copy, Clone, PartialEq, Debug)]
 pub struct Rational {
     pub numerator: u64,
     pub denominator: u64,
 }
 
+impl PartialOrd for Rational {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        (self.numerator as f64 / self.denominator as f64)
+            .partial_cmp(&(other.numerator as f64 / other.denominator as f64))
```
Comment (ruuda) [inline program/src/token.rs:34]: This comparison is lossy, `f64` cannot represent every `u64` exactly. This can lead to incorrect results:

```rust
#[test]
fn test_rational_cmp() {
    // Construct x and y such that x < y.
    let x = Rational {
        numerator: 1 << 53,
        denominator: 1,
    };
    let y = Rational {
        numerator: x.numerator + 1,
        denominator: x.denominator,
    };
    assert_eq!(x.partial_cmp(&y), Some(std::cmp::Ordering::Less));
}
```
```
thread 'token::test_rational_cmp' panicked at 'assertion failed: `(left == right)`
  left: `Some(Equal)`,
 right: `Some(Less)`', program/src/token.rs:49:5
```

We can fix this by avoiding floats. Observe that
```
   a/b < x/y
=> a   < bx/y    if b > 0
=> ay  < bx      if y > 0
```
We only work with non-negative denominators so the comparison flipping is not an issue, we only have to check to avoid division by zero:

```rust
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        if self.denominator == 0 || other.denominator == 0 {
            None
        } else {
            let x = self.numerator as u128 * other.denominator as u128;
            let y = other.numerator as u128 * self.denominator as u128;
            Some(x.cmp(&y))
        }
    }
```

The multiplications don't overflow because we widen to `u128`.

Please also add tests for other cases (like division by zero).

File: cli/src/daemon.rs [hunk 7]
```diff
@@ -133,6 +136,9 @@ impl MaintenanceMetrics {
                 self.transactions_unstake_from_inactive_validator += 1
```
[the remark this answers is not in the document]
Comment (enriquefynn) [reply cli/src/daemon.rs:136]: It is already there, below `RemoveValidator`.

File: cli/src/maintenance.rs [hunk 8]
```diff
@@ -836,6 +870,53 @@ impl SolidoState {
...
+
+        let (validator_index, unstake_amount) = lido::balance::get_unstake_validator_index(
+            &self.solido.validators,
+            &targets,
+            SolidoState::UNBALANCE_THRESHOLD,
+        )?;
+        let validator = &self.solido.validators.entries[validator_index];
+        let stake_account = &self.validator_stake_accounts[validator_index][0];
+
+        let maximum_unstake = (validator.entry.effective_stake_balance()
```
[the remark this answers is not in the document]
Comment (enriquefynn) [reply cli/src/maintenance.rs:894]: This is the maximum that can be unstaked from a validator, but afterwards we get the minimum from the stake account, this value and what should be unstaked.

File: cli/src/maintenance.rs [hunk 9]
```diff
@@ -836,6 +870,53 @@ impl SolidoState {
...
+            .expect("Validator should always have the minimum amount.");
+        // Get the maximum that can be unstaked from the stake account.  The
+        // minimum amongst the value to be unstaked, how much is in the stake
+        // account and the maximum that can be unstaked from the validator.
+        let amount = unstake_amount
+            .min(stake_account.1.balance.total())
+            .min(maximum_unstake);
+        if amount == Lamports(0) {
+            return None;
+        }
```
[the remark this answers is not in the document]
Comment (enriquefynn) [reply cli/src/maintenance.rs:905]: This would be always the case because of the comment above.

[the remark this answers is not in the document]
Comment (enriquefynn) [reply cli/src/maintenance.rs:905]: Ohh... I though the `MINIMUM_STAKE_ACCOUNT_BALANCE` was per validator's balance, otherwise we should keep track of the minimum to be rent exempt.

File: cli/src/maintenance.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (enriquefynn) [reply cli/src/maintenance.rs:894]: You're right 👍, the balance is per stake account and not per validator's balance

File: program/src/token.rs (hunk 6 above)
Comment (enriquefynn) [reply program/src/token.rs:34]: Thanks for the tests and more correct implementation, I'll change it and do more tests 👍

File: cli/src/daemon.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ruuda) [reply cli/src/daemon.rs:136]: No it’s not there, I mean here: https://github.com/ChorusOne/solido/blob/64509d59f7687470ef00f1d42fb5a8d09b91bfe5/cli/src/daemon.rs#L90-L107

File: cli/src/maintenance.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ruuda) [reply cli/src/maintenance.rs:905]: The reason we have the minimum was to work around the Solana bug, where it’s not possible to merge stake accounts if they contain to little SOL.

File: program/src/balance.rs [hunk 10]
```diff
@@ -102,6 +102,63 @@ pub fn get_target_balance(
...
+                Rational {
+                    numerator: target_difference,
+                    denominator: target.0,
+                } >= threshold
+            });
+
+    let (idx, validator) = validators
+        .entries
+        .iter()
+        .enumerate()
```
Comment (ruuda) [inline program/src/balance.rs:138]: We could zip it with `target_balance` and avoid the indexing:

```rust
validators
    .entries
    .iter()
    .zip(&target_balance)
    .max_by_key(|(validator, target)| ...)
```

File: program/src/token.rs [hunk 11]
```diff
@@ -257,4 +269,36 @@ pub mod test {
...
+    fn test_division_by_zero_cmp() {
+        let x = Rational {
+            numerator: 1,
+            denominator: 0,
+        };
+        let y = Rational {
+            numerator: x.numerator,
+            denominator: x.denominator + 1,
+        };
+        assert_eq!(x.partial_cmp(&y), None);
```
Comment (ruuda) [inline program/src/token.rs:297]: Please also test the reverse case of `y.partial_cmp(&x)`, that’s one of those places where a subtle copy-paste bug might hide. In fact, let’s also add the reverse case in the `test_rational_cmp` test above.

**Review by ruuda — COMMENTED** [review]
Looks almost good to me, I have a few small comments left.

File: program/src/balance.rs (hunk 10 above)
Comment (enriquefynn) [reply program/src/balance.rs:138]: Yeah, but we still need to return the index if we want to later get the stake account associated with the validator. We could pass it on the function, but I think it's not necessary.

Comment (ruuda) [reply program/src/balance.rs:138]: Fair point, never mind then.

**Review by ruuda — APPROVED** [review]
LGTM :shipit: :rocket:
